datafusion/core/src/datasource/mod.rs

```diff
@@ -175,3 +182,87 @@ fn get_col_stats(
+pub(crate) fn try_merge_schemas(
+    schemas: impl IntoIterator<Item = Schema>,
+    merge_compatible_types: bool,
+) -> Result<Schema> {
+    let mut metadata = HashMap::new();
+    let mut fields: Vec<Field> = vec![];
+    for schema in schemas {
+        for (key, value) in &schema.metadata {
+            if let Some(old_val) = metadata.get(key) {
+                if old_val != value {
+                    return Err(DataFusionError::ArrowError(ArrowError::SchemaError(
+                        format!(
+                            "Fail to merge schema due to conflicting metadata. \
+                                     Key '{}' has different values '{}' and '{}'",
+                            key, old_val, value
+                        ),
+                    )));
+                }
+            }
+            metadata.insert(key.to_owned(), value.to_owned());
+        }
+        for field in &schema.fields {
+            if let Some((i, merge_field)) =
+                fields.iter().find_position(|f| f.name() == field.name())
+            {
+                if merge_field.data_type() != field.data_type() {
+                    if let Some(new_type) =
+                        get_wider_type(merge_field.data_type(), field.data_type())
+                    {
+                        if &new_type != merge_field.data_type() {
+                            if merge_compatible_types {
+                                fields[i] = merge_field.clone().with_data_type(new_type);
+                            } else {
+                                return Err(DataFusionError::Execution(format!(
+                                    "Schema merge failed due to different, but compatible, data types for field '{}' ({} vs {}). \
+                                    Set '{}=true' (or call with_coerce_types(true) on reader options) \
+                                    to enable merging this field",
+                                    field.name(),
+                                    merge_field.data_type(),
+                                    field.data_type(),
+                                    OPT_FILE_FORMAT_COERCE_TYPES
+                                )));
+                            }
+                        }
+                    } else {
+                        return Err(DataFusionError::SchemaError(SchemaMergeError {
+                            field_name: field.name().to_owned(),
+                            data_types: (
+                                field.data_type().clone(),
+                                merge_field.data_type().clone(),
+                            ),
+                        }));
+                    }
+                }
+            } else {
+                // first time seeing this field
+                fields.push(field.clone());
+            }
+        }
+    }
+    Ok(Schema::new_with_metadata(fields, metadata))
+}
+
+pub(crate) fn get_wider_type(t1: &DataType, t2: &DataType) -> Option<DataType> {
```

Comment:
I would recommend using `comparison_coercion` here as I think it is a superset of these rules and will remain consistent with the rest of the codebase over time

https://github.com/apache/arrow-datafusion/blob/master/datafusion/expr/src/type_coercion/binary.rs#L185`

Reply:
There are also some popular coercions (like UTF8 -> Int*) that are not covered here